qutebrowser/browser/mhtml.py

```diff
@@ -0,0 +1,534 @@
+                             item.fileobj.getvalue(), mime, encode)
+        item.fileobj.actual_close()
+        if self.pending_downloads:
+            return
+        self.finish_file()
+
+    def error(self, url, item, *_args):
+        """Callback when a download error occurred.
+
+        Args:
+            url: The orignal url of the asset as QUrl.
+            item: The DownloadItem given by the DownloadManager.
+        """
+        try:
+            self.pending_downloads.remove((url, item))
+        except KeyError:
+            # This might happen if .collect_zombies() calls .finished() and the
+            # error handler will be called after .collect_zombies
+            log.downloads.debug("Oops! Download already gone: %s", item)
+            return
+        item.fileobj.actual_close()
+        # Add a stub file, see comment in .fetch_url() for more information
+        self.writer.add_file(urlutils.encoded_url(url), b'')
+        if self.pending_downloads:
+            return
+        self.finish_file()
+
+    def finish_file(self):
+        """Save the file to the filename given in __init__."""
+        if self._finished:
+            log.downloads.debug("finish_file called twice, ignored!")
+            return
+        self._finished = True
+        log.downloads.debug("All assets downloaded, ready to finish off!")
+        try:
+            file_output = open(self.dest, 'wb')
+        except OSError as error:
+            message.error(self._win_id,
+                          "Could not save file: {}".format(error))
+            return
+        else:
+            # Yes, ugly nesting of try-blocks, but necessary because there's
+            # no way to let the try block only affect the with-statement-
+            # entry-code and not the with-statement-body.
+            try:
+                self.writer.write_to(file_output)
+            finally:
+                file_output.close()
+        log.downloads.debug("File successfully written.")
+        message.info(self._win_id, "Page saved as {}".format(self.dest))
+
+    def collect_zombies(self):
+        """Collect done downloads and add their data to the MHTML file.
+
+        This is needed if a download finishes before attaching its
+        finished signal.
+        """
+        items = set((url, item) for url, item in self.pending_downloads
+                    if item.done)
+        log.downloads.debug("Zombie downloads: %s", items)
+        for url, item in items:
+            self.finished(url, item)
+
+
+class _NoCloseBytesIO(io.BytesIO):  # pylint: disable=no-init
+
+    """BytesIO that can't be .closed().
+
+    This is needed to prevent the DownloadManager from closing the stream, thus
+    discarding the data.
+    """
+
+    def close(self):
+        """Do nothing."""
+        pass
+
+    def actual_close(self):
+        """Close the stream."""
+        super().close()
+
+
+def _start_download(dest, win_id, tab_id):
```

Comment:
Why not pass the `web_view` instead of `tab_id` here as you need it in `start_download_checked` anyways?